libsplinter/src/network/peer_manager/mod.rs

```diff
@@ -0,0 +1,1040 @@
+        }
+    }
+
+    pub fn shutdown_and_wait(self) {
+        if let Some(sh) = self.shutdown_handle.clone() {
+            sh.shutdown();
+        } else {
+            return;
+        }
+
+        self.await_shutdown();
+    }
+}
+
+#[derive(Clone)]
+pub struct ShutdownHandle {
+    sender: Sender<PeerManagerMessage>,
+}
+
+impl ShutdownHandle {
+    pub fn shutdown(&self) {
+        if self.sender.send(PeerManagerMessage::Shutdown).is_err() {
+            warn!("Connection manager is no longer running");
+        }
+    }
+}
+
+fn handle_request(
+    request: PeerManagerRequest,
+    connector: Connector,
+    peers: &mut PeerMap,
+    peer_remover: &PeerRemover,
+    ref_map: &mut RefMap,
+) {
+    match request {
+        PeerManagerRequest::AddPeer {
+            peer_id,
+            endpoints,
+            sender,
+        } => {
+            if sender
+                .send(add_peer(
+                    peer_id,
+                    endpoints,
+                    connector,
+                    peers,
+                    peer_remover,
+                    ref_map,
+                ))
+                .is_err()
+            {
+                warn!("connector dropped before receiving result of adding peer");
+            }
+        }
+        PeerManagerRequest::UpdatePeer {
+            old_peer_id,
+            new_peer_id,
+            sender,
+        } => {
+            if sender
+                .send(update_peer(old_peer_id, new_peer_id, peers, ref_map))
+                .is_err()
+            {
+                warn!("connector dropped before receiving result of updating peer");
+            }
+        }
+        PeerManagerRequest::RemovePeer { peer_id, sender } => {
+            if sender
+                .send(remove_peer(peer_id, connector, peers, ref_map))
+                .is_err()
+            {
+                warn!("connector dropped before receiving result of removing peer");
+            }
+        }
+        PeerManagerRequest::ListPeers { sender } => {
+            if sender.send(Ok(peers.peer_ids())).is_err() {
+                warn!("connector dropped before receiving result of list peers");
+            }
+        }
+    };
+}
+
+fn add_peer(
+    peer_id: String,
+    endpoints: Vec<String>,
+    connector: Connector,
+    peers: &mut PeerMap,
+    peer_remover: &PeerRemover,
+    ref_map: &mut RefMap,
+) -> Result<PeerRef, PeerRefAddError> {
+    let new_ref_count = ref_map.add_ref(peer_id.to_string());
+
+    // if this is not a new peer, return sucess
+    if new_ref_count > 1 {
+        let peer_ref = PeerRef::new(peer_id, peer_remover.clone());
+        return Ok(peer_ref);
+    };
+
+    debug!("Attempting to peer with {}", peer_id);
+    // If new, try to create a connection
+    for endpoint in endpoints.iter() {
+        match connector.request_connection(&endpoint) {
+            Ok(()) => {
+                debug!("Peered with {}: {}", peer_id, endpoint);
```

Comment:
Isn't this more accurately "Peer {} connected via {}"

Reply:
updated